File: Days 61 - 70/Day 63/MatrixContainingWords.cpp
#include <array>
#include <iostream>
#include <string>

template <unsigned int Size>
bool ContainsWord(const std::array<std::array<char, Size>, Size>& matrix, const std::string& target)
{
	for (unsigned int i = 0; i < Size; i++)
	{
		std::string horizontal;

		for (const char letter : matrix[i])
		{
			horizontal += letter;
		}

		std::string vertical;

		for (unsigned int j = 0; j < Size; j++)
		{
			vertical += matrix[j][i];
		}

		if (horizontal == target || vertical == target)
		{
			return true;
		}
	}

	return false;
}

int main(int argc, char* argv[])
{
	const std::array<std::array<char, 4>, 4> matrix
	{
		std::array<char, 4> { 'F', 'A', 'C', 'I' },
		std::array<char, 4> { 'O', 'B', 'Q', 'P' },
		std::array<char, 4> { 'A', 'N', 'O', 'B' },
		std::array<char, 4> { 'M', 'A', 'S', 'S' },
	};

	std::cout << std::boolalpha;
	std::cout << ContainsWord(matrix, "FOAM") << "\n";
	std::cout << ContainsWord(matrix, "TEST") << "\n";
	std::cout << ContainsWord(matrix, "MASS") << "\n";

	std::cin.get();

	return 0;
}

File: Days 61 - 70/Day 63/MatrixContainingWords.cs
using System;

namespace DailyCodingProblem
{
	internal class Day63
	{
		private static int Main(string[] args)
		{
			char[,] matrix = new char[,]
			{
				{ 'F', 'A', 'C', 'I' },
				{ 'O', 'B', 'Q', 'P' },
				{ 'A', 'N', 'O', 'B' },
				{ 'M', 'A', 'S', 'S' },
			};

			Console.WriteLine(ContainsWord(matrix, "FOAM"));
			Console.WriteLine(ContainsWord(matrix, "TEST"));
			Console.WriteLine(ContainsWord(matrix, "MASS"));

			Console.ReadLine();

			return 0;
		}

		private static bool ContainsWord(char[,] letterMatrix, string word)
		{
			int width = letterMatrix.GetLength(1);
			int height = letterMatrix.GetLength(0);

			for (int i = 0; i < Math.Max(width, height); i++)
			{
				string horizonal = "";

				if (i < height)
				{
					for (int j = 0; j < width; j++)
					{
						horizonal += letterMatrix[i, j];
					}
				}

				string vertical = "";

				if (i < width)
				{
					for (int j = 0; j < height; j++)
					{
						vertical += letterMatrix[j, i];
					}
				}

				if (horizonal == word || vertical == word)
				{
					return true;
				}
			}

			return false;
		}
	}
}